File: day7.py
from array import array


class Dir:
    def __init__(self) -> None:
        self.children: dict = {}
        self.size: int = 0

    def is_dir(self) -> bool:
        return self.size == 0

    def __str__(self) -> str:
        return self.children.__str__() + " " + str(self.size)


def get_dirs():
    with open('inputs/day7.txt') as f:
        f.readline() # first line is irrelevant
        stack = []
        curr = Dir()
        for line in f.readlines():
            tokens = line.strip().split()
            if tokens[0] == '$':
                if tokens[1] == 'cd':
                    if tokens[2] == '..':
                        curr = stack.pop()
                    else:
                        stack.append(curr)
                        if tokens[2] in curr.children:
                            curr = curr.children[tokens[2]]
                        else:
                            next = Dir()
                            curr.children[tokens[2]] = next
                            curr = next
                else:
                    # this is an ls, do nothing
                    pass
            else:
                # we are currently in an ls
                if tokens[0] == 'dir':
                    if tokens[1] not in curr.children:
                        curr.children[tokens[1]] = Dir()
                else:
                    dir_file = Dir()
                    dir_file.size = int(tokens[0])
                    curr.children[tokens[1]] = dir_file
        return stack[0] if len(stack) > 0 else curr



def part1():
    top = get_dirs()

    def calculate_size(dir: Dir, sizes: array) -> int:
        size = dir.size
        for d in dir.children.values():
            size += calculate_size(d, sizes)

        if dir.is_dir() and size < 100_000:
            sizes.append(size)
        return size

    sizes = []
    calculate_size(top, sizes)
    print(sum(sizes))



def part2():
    top = get_dirs()

    def calculate_size(dir: Dir, sizes: array) -> int:
        size = dir.size
        for d in dir.children.values():
            size += calculate_size(d, sizes)

        if dir.is_dir():
            sizes.append(size)
        return size

    sizes = []
    total_size = calculate_size(top, sizes)

    unused_space = 70_000_000 - total_size
    space_to_free = 30_000_000 - unused_space

    closest = float('inf')
    for size in sizes:
        if size >= space_to_free and abs(space_to_free - size) < abs(space_to_free - closest):
            closest = size

    print(closest)


if __name__ == "__main__":
    part1()
    part2()

File: go/day7.go
package main

import (
	"fmt"
	"goaoc22/utils"
	"math"
	"strconv"
	"strings"
)

type Directory struct {
	size int
	children map[string]*Directory
}

func (dir *Directory) isDir() bool {
	return dir.size == 0
}

func MakeDir() *Directory {
	return &Directory{
		size: 0,
		children: map[string]*Directory{},
	}
}

func getDirs() *Directory {
	lines := utils.GetLines("../inputs/day7.txt")
	stack := []*Directory{}
	curr := MakeDir()

	for i := 1; i < len(lines); i++ {
		tokens := strings.Split(lines[i], " ")
		if tokens[0] == "$" {
			if tokens[1] == "cd" {
				if tokens[2] == ".." {
					curr = stack[len(stack) - 1]
					stack = stack[:len(stack) - 1]
				} else {
					stack = append(stack, curr)
					if _, ok := curr.children[tokens[2]]; ok {
						curr = curr.children[tokens[2]]
					} else {
						next := MakeDir()
						curr.children[tokens[2]] = next
						curr = next
					}
				}
			}
		} else {
			if tokens[0] == "dir" {
				if _, ok := curr.children[tokens[1]]; !ok {
					curr.children[tokens[1]] = MakeDir()
				}
			} else {
				file := MakeDir()
				size, err := strconv.Atoi(tokens[0])
				if err != nil {
					fmt.Println(err)
					panic(1)
				}
				file.size = size
				curr.children[tokens[1]] = file
			}
		}
	}

	return stack[0]
}

func calculateSize(dir *Directory, sizes *[]int) int {
	size := dir.size
	for _, d := range dir.children {
		size += calculateSize(d, sizes)
	}

	if dir.isDir() {
		*sizes = append(*sizes, size)
	}
	return size
}

func Day7() {
	sizes := []int{}
	totalSize := calculateSize(getDirs(), &sizes)
	unusedSpace := 70_000_000 - totalSize
	spaceToFree := 30_000_000 - unusedSpace

	closest := math.MaxUint32
	for _, size := range sizes {
		if size >= spaceToFree && utils.AbsInt(spaceToFree - size) < utils.AbsInt(spaceToFree - closest) {
			closest = size
		}
	}

	fmt.Println(closest)
}